.deb + .rpm via FPM

bernardoaraujor opened the pull request:
# Proposal

By using [FPM](https://fpm.readthedocs.io/en/latest/intro.html) and a shell script (`create-packages.sh`), we propose the automated creation of `nimbus_X.Y.Z_amd64.deb` and `nimbus_X.Y.Z_amd64.rpm` during release.

The script can be triggered by the [`release`](https://github.com/status-im/nimbus-eth2/blob/stable/.github/workflows/release.yml)  Workflow (not yet implemented).

# Considerations

Given the fact that we don't assume any `Nim` instalation  on the target system as a runtime pre-requisite, the package footprint differs from the current Tarball distribution.

Therefore, the underlying logic of the `run*.sh` scripts had to be adapted.

Here's a layout of the package image:
```
├── etc
│   └── default
│       └── nimbus.env
├── usr
│   └── bin
│   │   ├── nimbus_beacon_node
│   │   ├── nimbus_cli
│   │   ├── nimbus_cli_db
│   │   ├── nimbus_inspector
│   │   ├── nimbus_launch
│   │   ├── nimbus_signing_process
│   │   └── nimbus_validator_client
│   └── lib
│       └── systemd
│           └── system
│               └── nimbus.service
└── var
    └── lib
        └── nimbus
```

where:
- `/usr/lib/systemd/system/nimbus.service` holds the SystemD Service Unit definition.
- `/etc/default/nimbus.env` holds environment variables for the SystemD Service.
- `/var/lib/nimbus` holds the runtime data of the SystemD Service (`WorkingDir`).
- `/usr/bin/nimbus_launch` holds logic for launching `Nimbus` with proper `NETWORK` and `WEB3_URL` environment variables (inspired by `scripts/run-beacon-node.sh`).

We should also avoid making assumptions in regards to executable naming under the target runtime.
Therefore, some executables from `build/` were renamed with a `nimbus_*` prefix during population of `/usr/bin`.

# Testing

- `.deb` was manually tested under Ubuntu 20.04
- `.rpm` was manually tested under Fedora 33

# Limitations

- `/usr/bin/nimbus_launch` does not check for the validity of `WEB3_URL`, which could potentially lead to crashes with the executable.

- Testing covered for runtime execution, both via SystemD and manual. However, no assertions over the Ethereum protocol were made.

- Github Actions trigger not yet implemented. In case we get a green light on the PR draft, I'll send another commit calling `create-packages.sh` from `.github/workflows/release.yml`

bernardoaraujor committed:
bootstrap create-packages.sh

stefantalpalaru commented on create-packages.sh:
```diff
@@ -0,0 +1,68 @@
+cp build/ncli $PKG_IMG_DIR/usr/bin/nimbus_cli
+cp build/ncli_db $PKG_IMG_DIR/usr/bin/nimbus_cli_db
+cp build/nimbus_validator_client $PKG_IMG_DIR/usr/bin
+cp build/inspector $PKG_IMG_DIR/usr/bin/nimbus_inspector
```
Too many binaries in here. We only need two - just like in the distributed tarballs.

bernardoaraujor replied:
f55b9dd089da5d1cc6529945684ff9460a221653

stefantalpalaru commented on create-packages.sh:
```diff
@@ -0,0 +1,68 @@
+  --after-install $PKG_IMG_DIR/postinst \
+  --before-remove $PKG_IMG_DIR/prerm \
+  --after-remove $PKG_IMG_DIR/postrm \
+  --license "Apache 2.0" \
```
Can we also specify the second license?

bernardoaraujor replied:
9187927424050660bd006e562401eae694d96887

stefantalpalaru commented on package_image/etc/default/nimbus.env:
```diff
@@ -0,0 +1,12 @@
+# DATA_DIR="/var/lib/nimbus/${DATA_DIR_NAME}"
+# BASE_P2P_PORT=9000
+# BASE_RPC_PORT=9190
+WEB3_URL=wss://mainnet.infura.io/ws/v3/8a287a9063074512bfefb1a52fa50c78
```
We can't publish the Infura API token for a single account, because it's guaranteed to run out of its monthly quota as soon as people start using it.

Also, are these env vars landing in the global namespace? We need something that's only sourced by the Nimbus-eth2 service, not any login shell.

bernardoaraujor replied:
If Nimbus will be available as a SystemD Service, there needs to be a way for the user to specify the Web3 provider.

There's different approaches to this problem, and people tend to be quite opinionated about it.
In order to give you a decent answer, I went down a rabbit whole and found some stuff I wasn't really aware of:
- http://0pointer.de/blog/projects/on-etc-sysinit.html
- https://unix.stackexchange.com/a/419061

To answer your question, the variables defined under `/etc/default/nimbus.env` do not land into global namespace.
But as you can see from the links above, I'm not really married to this solution anymore.

As suggested [here](https://github.com/status-im/nimbus-eth2/pull/2691#discussion_r662618525), I could get rid of `/usr/bin/nimbus_launch` and replace it with the following in the Service Unit file:

```
...
Environment=NETWORK="mainnet"
Environment=NBC_BINARY=/usr/bin/nimbus_beacon_node
Environment=NODE_ID=0
Environment=DATA_DIR_NAME="shared_${NETWORK}_${NODE_ID}"
Environment=DATA_DIR="/var/lib/nimbus/${DATA_DIR_NAME}"
Environment=BASE_P2P_PORT=9000
Environment=BASE_RPC_PORT=9190
Environment=WEB3_URL=""
...
ExecStart="${NBC_BINARY} --network=${NETWORK} --data-dir=${DATA_DIR} --log-file=${DATA_DIR}/nbc_bn_$(date +\"%Y%m%d%H%M%S\").log --tcp-port=$(( ${BASE_P2P_PORT} + ${NODE_ID} )) --udp-port=$(( ${BASE_P2P_PORT} ${NODE_ID} )) --rpc --rpc-port=$(( ${BASE_RPC_PORT} +${NODE_ID} )) --web3-url=${WEB3_URL}"
```

However we'd have to instruct the end-user to set `WEB3_URL` before starting the service. In case they forgot, they would have to rely on the following to figure out what's going on:

```
$ sudo journalctl -u nimbus | tail
Jul 02 01:11:22 ubuntu01 systemd[1]: Started Nimbus.
Jul 02 01:11:22 ubuntu01 nimbus[277]: mainnet
Jul 02 01:11:22 ubuntu01 nimbus[277]: Please enter a Web3 provider URL:
Jul 02 01:11:22 ubuntu01 systemd[1]: nimbus.service: Main process exited, code=exited, status=1/FAILURE
Jul 02 01:11:22 ubuntu01 systemd[1]: nimbus.service: Failed with result 'exit-code'.
Jul 02 01:11:22 ubuntu01 systemd[1]: nimbus.service: Scheduled restart job, restart counter is at 5.
Jul 02 01:11:22 ubuntu01 systemd[1]: Stopped Nimbus.
Jul 02 01:11:22 ubuntu01 systemd[1]: nimbus.service: Start request repeated too quickly.
Jul 02 01:11:22 ubuntu01 systemd[1]: nimbus.service: Failed with result 'exit-code'.
Jul 02 01:11:22 ubuntu01 systemd[1]: Failed to start Nimbus.
```

bernardoaraujor replied:
in progress... I'll push a commit soon

arnetheduck replied:
We should probably move the check for web3-url to the beacon node binary instead - basically, it's required when running a validator and optional otherwise, so the beacon  node can make that check - at the same time, when a web3 is provided, we should always use it to sync eth1 - right now, we don't, unless there's a validator - this satisfies https://github.com/status-im/nimbus-eth2/issues/2685 and overall makes the whole configuration more simple and consistent.

These changes are not necessarily for this PR - we can assume that they will happen though and model this PR after that behavior - ping @zah @mratsim

zah replied:
Our long-term vision is that users will be installing Nimbus with a command such as `apt install nimbus`  and this would automatically create a service that connects and monitors Eth2 mainnet. The user would then use additional commands to attach validators to the running service and for this to be possible, the web3 parameter needs to be specified (it's partially optional for the pure monitoring function). Technically, the web3 parameter is specified per network, so the `env` configuration file can be stored in the same directory where the network data resides (i.e. the `--data-dir` folder).

Some users will be interested in joining multiple networks, but for now we can handle this by requiring the user to manually create an additional SystemD unit file launching a separate beacon node. 

What I think is important for the installer to do right is to configure properly the automatically installed default service. To solve this, we can introduce an interactive FPM installation script that will use an utility like [dialog](https://linux.die.net/man/1/dialog) to create some prompts asking the user to select a network (defaults to `mainnet`) and to enter a web3 URL with prompts such as these:

https://docs.openstack.org/ocata/cs/install-guide-debconf/_images/dbconfig-common_2_db-types.png

https://r00t4bl3.com/uploads/attachment-screenshot-from-2020-09-17-07-36-53-a34ba9c70a80b1265e06d3bfa6728e1c.png

It should be possible to run the installer in an unattended fashion though where these values are specified in a wrapper script through some environment variables.

The web3-url dialog is likely to feature a lengthy explanation for the possible values that our team is going to be working on in separate PRs.

bernardoaraujor replied:
@zah to be clear, I later found out that using `.env` files and `EnvironmentFile=` is frowned upon in the SystemD community (check the links in my comment above).
So I'm replacing the environment variables as `Environment=` definitions in the Unit file.

In summary, all Nimbus configurations are being set at `/usr/lib/systemd/system/nimbus.service`.
In the documentation we should make the users aware of that.

And in case the user desires to have additional services for different networks, they should do so by creating new Unit files like:
- `/usr/lib/systemd/system/nimbus_mainnet.service`
- `/usr/lib/systemd/system/nimbus_pyrmont.service`
- `/usr/lib/systemd/system/nimbus_prater.service`

And setting `NETWORK` + `WEB3_URL` accordingly.

stefantalpalaru commented on package_image/postinst:
```diff
@@ -0,0 +1,6 @@
+#!/bin/sh
+
+set -e
+chmod 777 /var/lib/nimbus
```
Way too permissive. Better let the user create the data dir instead.

(Or find out how to create a package-specific user:group, make that the dir owner and make sure the service runs under it)

bernardoaraujor replied:
2aace12a3f90d5b256ce6d2197f1a37b23451a9b

stefantalpalaru commented on package_image/postinst:
```diff
@@ -0,0 +1,6 @@
+set -e
+chmod 777 /var/lib/nimbus
+
+exit 0
```
Why? The default exit value is that of the last command.

stefantalpalaru commented on package_image/prerm:
```diff
@@ -0,0 +1,11 @@
+if type systemctl >/dev/null 2>/dev/null; then
+	if [ -z "$D" ]; then
+		systemctl stop nimbus.service
+		systemctl disable nimbus.service
```
Will this script also run for package updates?

bernardoaraujor replied:
Based on [this](https://www.debian.org/doc/debian-policy/ch-maintainerscripts.html#summary-of-ways-maintainer-scripts-are-called) and [this](https://www.debian.org/doc/debian-policy/ch-maintainerscripts.html#details-of-unpack-phase-of-installation-or-upgrade), I believe the answer is yes.

stefantalpalaru replied:
Then we can't mess with the service in it. What's in "D"?

stefantalpalaru commented on package_image/usr/bin/nimbus_launch:
```diff
@@ -0,0 +1,81 @@
+fi
+
+# Allow the binary to receive signals directly.
+exec ${WINPTY} ${NBC_BINARY} \
```
A script that no developer will use is not going to be maintained.

It's probably easier to move the command in the Systemd service file itself.

bernardoaraujor replied:
in progress... I'll push a commit soon

bernardoaraujor committed:
remove unnecessary binaries from create-packages.sh

bernardoaraujor committed:
add MIT license to create-packages.sh

bernardoaraujor committed:
rm unnecessary exit status from package_image scripts

bernardoaraujor committed:
add nimbus user and group on package install

stefantalpalaru commented:
BTW, this is how I do it in Gentoo:

- config file: https://github.com/stefantalpalaru/gentoo-overlay/blob/master/net-p2p/nimbus-eth2/files/nimbus-eth2.conf
- init script: https://github.com/stefantalpalaru/gentoo-overlay/blob/master/net-p2p/nimbus-eth2/files/nimbus-eth2.init

bernardoaraujor commented:
as suggested by @zah I'm introducing [dialog](https://linux.die.net/man/1/dialog) to allow the user to choose network and Web3 provider.
That's being implemented into the `postinst` script, which runs every time after a fresh installation or upgrade.

I'm adding some logical safeguards, and this question came to mind:

What happens if `nimbus_beacon_node` starts with an empty `--web3-url=` arg?
It doesn't seem to prevent the executable from starting, as in the case of an unrecognized or empty `--network` parameter.

I feel we should still warn the user in case no `WEB3_URL` is provided after installation via `dialog`.
Something like:

```
echo "You didn't provide any Eth1 Web3 Provider.
xxx
You should fix this by editing /usr/lib/systemd/system/nimbus.service and setting proper value on WEB3_URL."
```

What should I write into `xxx` to give the user an accurate notion of how Nimbus is going to behave?

bernardoaraujor committed:
rm nimbus.env and nimbus_launch; improve systemd unit

stefantalpalaru commented:
> I'm introducing dialog to allow the user to choose network and Web3 provider

Interactive package installation/upgrade is a bad idea. It's better to have the service refuse to start until a text configuration file is being edited by hand.

zah commented:
> Interactive package installation/upgrade is a bad idea. It's better to have the service refuse to start until a text configuration file is being edited by hand.

Interactive prompts during package installation are not unprecedented (I've seen them in mailservers, databases and other packages with comparable complexity to Nimbus). What is so bad about them?

stefantalpalaru commented:
> What is so bad about them?

They sabotage attempts at automation. See how I have to set "ENV DEBIAN_FRONTEND=noninteractive" here:
https://github.com/status-im/nimbus-eth2/blob/44f652f704f14c67e648f2b14dee652d44b858c5/docker/dist/base_image/Dockerfile.amd64#L13

If you come up with ways to go around those established ways to handle config files, not even that env var will save me from having to deal with interactive prompts inside scripts.

zah commented:
My original suggestion featured the following:

> It should be possible to run the installer in an unattended fashion though where these values are specified in a wrapper script through some environment variables.

bernardoaraujor committed:
fix prerm

bernardoaraujor committed:
fix postrm userdel

bernardoaraujor commented:
I came up with a compromise strategy for the constraints mentioned by @stefantalpalaru and @zah:

The `postinst` script defines a `dialog_user()` function.
The script does a `read` and expects `Y/Yes/y/yes` or `N/No/n/no`, and only calls `dialog` iff the input is positive.

Any automation script can call the [yes](https://man7.org/linux/man-pages/man1/yes.1.html) utility to avoid the usage of `dialog`, such as:
```
$ yes N | dpkg -i nimbus_0.2.0_amd64.deb
```

bernardoaraujor committed:
postinst asks about dialog

zah commented:
Instead of merely preventing user input, the unattended install must be able to provide specific values such as `NIMBUS_NETWORK` and `NIMBUS_WEB3_URL` that will end up in the created SystemD configuration.

bernardoaraujor commented:
> Instead of merely preventing user input, the unattended install must be able to provide specific values such as `NIMBUS_NETWORK` and `NIMBUS_WEB3_URL` that will end up in the created SystemD configuration.

`NETWORK` is set to `mainnet` as default on unattended install (check [here](https://github.com/bernardoaraujor/nimbus-eth2/blob/fb157bd0902d096d70099c4571a35a1f07c26d17/package_image/postinst#L53)).
But `WEB3_URL` is non-trivial.

@stefantalpalaru pointed [here](https://github.com/status-im/nimbus-eth2/pull/2691#discussion_r662615022) that we can't publish an Infura API token for a single account, because it's guaranteed to run out of its monthly quota as soon as people start using it.

Is there some URL that could be shared?

Keep in mind that unattended install prints out the following:
```
echo "Warning: you must manually edit /usr/lib/systemd/system/nimbus.service and setting proper values on NETWORK and WEB3_URL."
```

zah commented:
What I mean is that the user should be able to pre-select the values entered interactively during the setup by providing them as environment variables:

```bash
export NIMBUS_WEB3_URL=https://localhost:8156
export NIMBUS_NETWORK=pyrmont
sudo dpkg -i nimbus.deb
```

bernardoaraujor committed:
create-packages.sh supported architectures

bernardoaraujor committed:
postinst NETWORK + WEB3_URL preset env vars

bernardoaraujor commented:
@zah I was able to achieve that only partially (64a6d7f21f5c285020dce7bf4a6554cb9ebca621)

The problem with `export` is that it's executed by `$USER`, while `dpkg` is executed by `root` (via `sudo`).
Therefore, `NETWORK` and `WEB3_URL` are not available when `dpkg` is called.

This, however, works:
```
$ sudo NETWORK=pyrmont WEB3_URL=https://localhost:8156 dpkg -i nimbus.deb
```

Also, note that I'm not using the `NIMBUS_*` prefix from your examples.
It wasn't used initially (nor in the scripts I took as reference), so I kept it out for consistency. Let me know whether I should rename the vars in order to use such prefix.

bernardoaraujor commented:
@zah Another issue I'd like to point out is the fact that using `dialog` introduces it a dependency.
For example, when I'm testing the package on virgin Ubuntu VMs, before I'm able to install it I need to do:
```
$ sudo add-apt-repository universe
$ sudo apt-get install dialog
```

A similar (but simpler) approach could be achieved using `read`.

Not as eye-candy as `dialog`, but reduces complexity on user side.

bernardoaraujor committed:
Merge branch 'unstable' of https://github.com/status-im/nimbus-eth2 into unstable

arnetheduck commented:
https://geth.ethereum.org/docs/install-and-build/installing-geth#install-on-ubuntu-via-ppas - geth has a PPA for these packages - should we somehow consider that, ie from what I can tell, they package each piece into a separate deb, then have a general `ethereum` package that installs everything.

In some ways, it would be nice to perhaps integrate the two, through there are also downsides (geth monoculture)

zah commented on create-packages.sh:
```diff
@@ -0,0 +1,74 @@
+fi
+
+PKG_VERSION=$2
+if ! [[ "$PKG_VERSION" =~ ^[0-9]+\.[0-9]+ ]]; then
```
Our versions have 3 components (e.g. 1.4.1)
You can get the version from the binary by executing `nimbus_beacon_node --version` or by creating a small `print_version.nims` module that imports `beacon_chain/version.nim` and prints any specific value you are interested in.

bernardoaraujor replied:
8f432208d3c18de8f17c046c30353e4d7af37e1f

bernardoaraujor replied:
841eccaac89509765aa597722087cd480ada7e95

zah commented on package_image/postinst:
```diff
@@ -0,0 +1,60 @@
+  fi
+
+  # check for Web3 Provider URL prefixes (https:// or wss://)
+  if [[ !( -z $WEB3_URL ) && !( $WEB3_URL =~ ^https\:\/\/.* || $WEB3_URL =~ ^wss\:\/\/.* ) ]]; then
```
It's OK for a the web3-ulr to start with `http://` or `ws://` (the non-secure versions of the protocols). This is likely to be the case when the user wants to pair Nimbus with a locally running Geth instance.

bernardoaraujor replied:
bbfc0328d96e87123eb57d7e28d46427e4578785

zah commented on package_image/postinst:
```diff
@@ -0,0 +1,60 @@
+}
+
+while true; do
+  read -p "Do you want to be assisted in the configuration of the Nimbus SystemD Service? y/n " yn
```
I don't think this prompt is necessary if we implement the ENV variables for the unattended setup. In other words, if the ENV variable is present, its respective dialog can be skipped.

bernardoaraujor replied:
a6bf0dfd5d53f2f7f6476a385f7a865300b38846

zah commented on package_image/postrm:
```diff
@@ -0,0 +1,6 @@
+set -e
+rm -rf /var/lib/nimbus
+rm -rf /usr/lib/systemd/system/nimbus.service
+userdel nimbus
```
`userdel` will fail if the user doesn't exist (perhaps it was deleted manually by the administrator). This in turn will fail the entire `postrm` script due to the `set -e` directive. I think it's better if we handle this case gracefully without an error.

bernardoaraujor replied:
afb0bb710275e09238bd1b4878bbfd83c905f2bd

zah commented on package_image/prerm:
```diff
@@ -0,0 +1,8 @@
+#!/bin/sh
+
+set -e
+if (systemctl is-active --quiet nimbus.service)
```
`systemctl is-active` seems to test whether the service is currently running. This is an adequate pre-condition for running `systemctl stop`, but it isn't for `systemctl disable`. If the service was enabled (i.e. schedule to start automatically upon booting), it should be disabled during the uninstall regardless of its current running state.

bernardoaraujor replied:
b8acd1deb5676c34b8793477cc8c831e75864903

zah replied:
The linked commit doesn't implement the solution I have suggested. `systemctl disable` should be executed if the service was enabled (to start automatically on boot up), but it's not currently running.

zah commented on package_image/usr/lib/systemd/system/nimbus.service:
```diff
@@ -0,0 +1,29 @@
+
+[Service]
+LimitNOFILE=4096
+StandardOutput=syslog
```
What will happen if we don't specify these? Would a wider system policy take place?

bernardoaraujor replied:
removed a few configs, allowing default values to take place
service didn't show significant behavior changes
fd702a12f415819978c29d8f0ce4f4ea95f409aa

zah replied:
The following removed options:

```
PrivateDevices=yes
PrivateTmp=yes
ProtectSystem=full
ProtectHome=yes
```

... look like security hardening measures that we'll be happy to accept (together will any other hardening measure we are compatible with). I expect that many of these measures are not enabled by default in SystemD due to backwards compatibility concerns and they have to be opted-in.

zah commented on package_image/usr/lib/systemd/system/nimbus.service:
```diff
@@ -0,0 +1,29 @@
+WorkingDirectory=/var/lib/nimbus
+TimeoutSec=1200
+Restart=always
+ExecStart=/bin/bash -c '/usr/bin/nimbus_beacon_node --network=${NETWORK} --data-dir=${DATA_DIR_PREFIX}/shared_${NETWORK}_${NODE_ID} --log-file=${DATA_DIR_PREFIX}/shared_${NETWORK}_${NODE_ID}/nbc_bn_$$(date +%%Y%%m%%d%%H%%M%%S).log --tcp-port=${TCP_PORT} --udp-port=${UDP_PORT} --rpc --rpc-port=${RPC_PORT} --web3-url=${WEB3_URL}'
```
The `--log-file` option is not recommended when running under SystemD bacause journald provides enough features for archiving and inspecting the produced logs. 

To be safe, it's always better to quote paths on the command line such as `--data-dir="X"`, because ENV vars such as `DATA_DIR_PREFIX` may contain spaces.

bernardoaraujor replied:
e11ae370a47a10648523645069464d9544dc0568
bfdd6802167475aae8684ca0093a9259b52ba38e

arnetheduck commented:
> But WEB3_URL is non-trivial.

Given that if you're running the node just to monitor, we don't need it, we could simply skip it for unattended installs, wdyt @zah ?

bernardoaraujor committed:
Merge branch 'unstable' of https://github.com/status-im/nimbus-eth2 into unstable

bernardoaraujor committed:
check binaries exist

bernardoaraujor committed:
awk for pkg version

bernardoaraujor committed:
allow http+ws for WEB3_URL on postinst dialog

bernardoaraujor committed:
use env vars  +  for dialog

bernardoaraujor committed:
check nimbus user on postinst and postrm

bernardoaraujor committed:
rm systemctl disable

bernardoaraujor committed:
rm --log-file option

bernardoaraujor committed:
add quotes to --data-dir path

bernardoaraujor committed:
rm extra /

bernardoaraujor committed:
rm unneeded systemd unit vars

bernardoaraujor commented:
If I build the package on my Debian 10 and install it on an Ubuntu 16.04 VM, the following happens when I  try to execute it:

```
$ nimbus_beacon_node 
nimbus_beacon_node: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by nimbus_beacon_node)
nimbus_beacon_node: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by nimbus_beacon_node)
```

That's because dynamic linking is introducing runtime dependencies that will not be necessarily met on every target, even in cases of slightly differing versions of system libraries such as `glibc`, as demonstrated above.

One possible way to mitigate this could be to use `ldd` to list the dynamic lib dependencies. However this approach presents many obstacles, namely:
- the list of dynamic lib dependencies would need to be generated dynamically by the script, since the static linking is determined by the build machine (CI in **most**, but not necessarily **all** cases).
- the output of `ldd` would need to be parsed properly in order to become a list of `-d <dependency>` options for `fpm`.
- to require specific versions of system libs such as `glibc` feels too radical and restrictive, leaving out many target candidates.

Another much simpler approach would be to just force static linking during the build.
I found `--app:staticlib` on the [Nim Compiler User Guide](https://nim-lang.github.io/Nim/nimc.html).

Should we automate that as an option in the `Makefile`?

stefantalpalaru commented:
It's not a good idea to statically link Glibc: https://stackoverflow.com/questions/57476533/why-is-statically-linking-glibc-discouraged

Now, newer Glibc versions are more optimised, but we want the binaries to run on more than the last distro version, so there's a trade-off here. The oldest distro we build on is Ubuntu-18.04 for ARM64, so we support Raspbian: https://github.com/status-im/nimbus-eth2/blob/07a1c5716b707e7244eb312bed68c92c1fb9508d/docker/dist/base_image/Dockerfile.arm64#L9-L11

Other platforms use Ubuntu-20.04.

bernardoaraujor committed:
improve build error msg

bernardoaraujor committed:
check fpm installation

bernardoaraujor committed:
check nimbus user via id command

bernardoaraujor committed:
simplify postinst conditionals

bernardoaraujor committed:
avoid fedora interactive postinst; fix fedora adduser

From https://docs.fedoraproject.org/en-US/Fedora_Draft_Documentation/0.1/html/RPM_Guide/ch09s04s05.html

"Warning
Do not try to write interactive scripts. Many users install RPMs automatically.
In such cases, or if the user runs a graphical RPM tool, any information your scripts output will be lost.
User input will not be available."
